### DISCUSSION TITLE
teoria relazionale - 10

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#77] [Esercizi]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/a15e9fa8-928b-4253-a3fa-ffc4f574816f)

**Comment by IonutCicio** [comment]
![Image (18)](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/a07b66be-e54d-48b5-a293-7341bbe7091d)

**Reply by IonutCicio** [reply]
Il codice a cui faccio riferimento l'ho messo qui #55

**Comment by IonutCicio** [comment]
$\mathcal{P}(R)$ = { $\varnothing$, A, B, C, D, E, H, I, AB, AC, AD, AE, AH, AI, BC, BD, BE, BH, BI, CD, CE, CH, CI, DE, DH, DI, EH, EI, HI, ABC, ABD, ABE, ABH, ABI, ACD, ACE, ACH, ACI, ADE, ADH, ADI, AEH, AEI, AHI, BCD, BCE, BCH, BCI, BDE, BDH, BDI, BEH, BEI, BHI, CDE, CDH, CDI, CEH, CEI, CHI, DEH, DEI, DHI, EHI, ABCD, ABCE, ABCH, ABCI, ABDE, ABDH, ABDI, ABEH, ABEI, ABHI, ACDE, ACDH, ACDI, ACEH, ACEI, ACHI, ADEH, ADEI, ADHI, AEHI, BCDE, BCDH, BCDI, BCEH, BCEI, BCHI, BDEH, BDEI, BDHI, BEHI, CDEH, CDEI, CDHI, CEHI, DEHI, ABCDE, ABCDH, ABCDI, ABCEH, ABCEI, ABCHI, ABDEH, ABDEI, ABDHI, ABEHI, ACDEH, ACDEI, ACDHI, ACEHI, ADEHI, BCDEH, BCDEI, BCDHI, BCEHI, BDEHI, CDEHI, ABCDEH, ABCDEI, ABCDHI, ABCEHI, ABDEHI, ACDEHI, BCDEHI, ABCDEHI}

Questo è l'insieme potenza di $R$, come facciamo a dire che le 4 sopra sono le sole e uniche chiavi se non sappiamo quante sono?

- $\varnothing$ non può essere chiave
- A, B, C, D, E, H e I, presi singolarmente, non sono chiavi
- H non compare mai a sinistra, quindi possiamo eliminare tutti i sottoinsiemi con H _(dato che non contribuirà mai ad una chiave)_

{AB, AC, AD, AE, AI, BC, BD, BE, BI, CD, CE, CI, DE, DI, EI, ABC, ABD, ABE, ABI, ACD, ACE, ACI, ADE, ADI, AEI, BCD, BCE, BCI, BDE, BDI, BEI, CDE, CDI, CEI, DEI, ABCD, ABCE, ABCI, ABDE, ABDI, ABEI, ACDE, ACDI, ACEI, ADEI, BCDE, BCDI, BCEI, BDEI, CDEI, ABCDE, ABCDI, ABCEI, ABDEI, ACDEI, BCDEI, ABCDEI}

- scopriamo abbastanza facilmente che AB è chiave, quindi possiamo eliminare tutte le superchiavi che contengono propriamente AB

{**AB**, AC, AD, AE, AI, BC, BD, BE, BI, CD, CE, CI, DE, DI, EI, ACD, ACE, ACI, ADE, ADI, AEI, BCD, BCE, BCI, BDE, BDI, BEI, CDE, CDI, CEI, DEI, ACDE, ACDI, ACEI, ADEI, BCDE, BCDI, BCEI, BDEI, CDEI, ACDEI, BCDEI}

- scopriamo che CD è chiave, quindi possiamo eliminare tutte le superchiavi che contengono propriamente CD

{**AB**, AC, AD, AE, AI, BC, BD, BE, BI, **CD**, CE, CI, DE, DI, EI, ACE, ACI, ADE, ADI, AEI, BCE, BCI, BDE, BDI, BEI, CEI, DEI, ACDE, ACEI, ADEI, BCEI, BDEI}

- scopriamo che DA è chiave, quindi possiamo eliminare tutte le superchiavi che contengono propriamente DA

{**AB**, AC, **AD**, AE, AI, BC, BD, BE, BI, **CD**, CE, CI, DE, DI, EI, ACE, ACI, AEI, BCE, BCI, BDE, BDI, BEI, CEI, DEI, ACEI BCEI, BDEI}

- scopriamo ✨ _magicamente_ ✨ che BC è chiave, ...

{**AB**, AC, **AD**, AE, AI, **BC**, BD, BE, BI, **CD**, CE, CI, DE, DI, EI, ACE, ACI, AEI, BDE, BDI, BEI, CEI, DEI, ACEI, BDEI}

A questo punto bisognerebbe procedere a calcolare le chiusure di quello che rimane, partendo dagli insiemi più grandi...

- BDEI non è chiave, quindi possiamo eliminare BD, BE, BI, DE, DI etc...

{**AB**, AC, **AD**, AE, AI, **BC**, BE, **CD**, CE, CI, ACE, ACI, AEI, BDE, CEI, ACEI}

- ACEI non è chiave

{**AB**, **AD**, **BC**, BE, **CD**, BDE}

- BDE non è chiave

{**AB**, **AD**, **BC**, **CD**}

Sono rimaste solo queste chiavi! Il problema è che al compito non possiamo permetterci di scrivere $2^6$ chiavi e cancellarle così, quindi proviamo a trovare un altro modo per eliminiare le possibili chiavi, partendo da 

{AB, AC, AD, AE, AI, BC, BD, BE, BI, CD, CE, CI, DE, DI, EI, ABC, ABD, ABE, ABI, ACD, ACE, ACI, ADE, ADI, AEI, BCD, BCE, BCI, BDE, BDI, BEI, CDE, CDI, CEI, DEI, ABCD, ABCE, ABCI, ABDE, ABDI, ABEI, ACDE, ACDI, ACEI, ADEI, BCDE, BCDI, BCEI, BDEI, CDEI, ABCDE, ABCDI, ABCEI, ABDEI, ACDEI, BCDEI, ABCDEI}

- ABCDE, ABCDI, ABCEI, ABDEI, ACDEI, BCDEI sono tutte superchiavi, non si può dire nulla
- ABCD, ABCE, ABCI, ABDE, ABDI, ABEI, ACDE, ACDI, ADEI, BCDE, BCDI, BCEI, CDEI sono superchiavi
- ACEI, BDEI non lo sono

etc...

Partire al contrario dai sottoinsiemi più grandi non è vantaggioso, perché si rivela che una gran parte sono superchiavi, e quindi bisognerebbe ricontrollare un tante chiusure.

> [!TIP]
> Un modo c'è!

Per scrivere tutte e 64 le combinazioni ci ho messo una prima volta $\approx$ 5 minuti, una seconda volta $\approx$ 3 minuti e 20; secondo me, quello che conviene fare se non conosciamo il numero preciso di chiavi, è scriversi le 64 combinazioni e fare l'eliminazione che ho fatto con il primo metodo _(sbarrando le non chiavi)_, altrimenti non si hanno modi certi per decretare che quelle sono le sole e uniche chiavi, se non con ragionamenti che richiederebbero più tempo rispetto a scrivere i sottoinsiemi e cancellare quelli non validi.

**Comment by daromacarlo** [comment]
![Immagine 2024-02-05 160835](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/5915b274-78da-4c52-b93c-c316cf6d7998)
![Immagine 2024-02-05 160852](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/0d9d2bd8-cb37-4df2-8bfe-f6f2672044eb)
![Immagine 2024-02-05 160910](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/89a93952-9ea7-457f-814f-5d93326279a9)
![Immagine 2024-02-05 160927](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/565ff973-2d81-4c1a-ad25-760d08896f87)

**Reply by daromacarlo** [reply]
anche BDEI non è chiave.
